### ISSUE TITLE
zfs send -R can not order clones

### ISSUE THREAD

**Issue by orivej** [zfsonlinux/zfs#4347]
I'm trying to replicate Docker on ZFS onto a new pool and it fails like this:
```
# zfs snapshot -r base/root/docker@new
# zfs send -RLe base/root/docker@new | zfs receive -v new/root/docker
receiving full stream of base/root/docker@new into new/root/docker@new
received 8.76MB stream in 1 seconds (8.76MB/sec)
cannot receive: local origin for clone new/root/docker/62b96e8f1ecc2ecbea76ba662379bc19c60aa655b38261042f6b0cafda71759b@333857985 does not exist

# zfs get -H -o value origin base/root/docker/62b96e8f1ecc2ecbea76ba662379bc19c60aa655b38261042f6b0cafda71759b
base/root/docker/f6e6c8078ba47d5b9544b741f46979e2d6ddd4bea830b63d01f4efe8c4ac6b43@977616200

# zfs list base/root/docker/f6e6c8078ba47d5b9544b741f46979e2d6ddd4bea830b63d01f4efe8c4ac6b43@977616200
NAME                                                                                          USED  AVAIL  REFER  MOUNTPOINT
base/root/docker/f6e6c8078ba47d5b9544b741f46979e2d6ddd4bea830b63d01f4efe8c4ac6b43@977616200     1K      -   149M  -

# zfs list -H -o name -r base/root/docker | xargs -n1 zfs get -H -o name,value origin
#gist below
```
https://gist.github.com/orivej/caaa9b7bf688249cd8db

This happens under Gentoo with zfs 0.6.5.4-r2, zfs-kmod 0.6.5.4-r1, spl 0.6.5.4-r1.

**Comment by orivej** [comment]
I sorted my list of docker filesystems, snapshots and clones into a script that recreates the same structure. (Commited to the same gist: https://gist.github.com/orivej/caaa9b7bf688249cd8db#file-zfs-recreate-sh)

Unfortunately I can not reproduce the original issue because ZFS hangs (`zfs list` is unkillable etc.) when I attempt a send/receive. (Now I am on zfs 0.6.5.3-r5, sys-fs/zfs-kmod 0.6.5.3-r4, sys-kernel/spl 0.6.5.3-r1 since 0.6.5.4 hung too often when I manipulated ZFS filesystems.) Do you have a complete script for reporting issues that formalizes what is written at http://zfsonlinux.org/faq.html#HowDoIReportProblems?

**Comment by orivej** [comment]
I checked that my `zfs-recreate.sh` works fine and allows subsequent recursive send/receive under Fedora 23. I have also upgraded to the latest versions from git without Gentoo-specific patches and was able to successfully send/receive my Docker filesystem hierarchy. I'm going to close this issue since it does not apply to the official zfsonlinux, but if needed, I could help investigate the issue with Ryan patches.
